libs/SalesforceSDK/src/com/salesforce/androidsdk/rest/ClientManager.java

```diff
@@ -379,8 +379,20 @@ public AccMgrAuthTokenProvider(ClientManager clientManager, String instanceUrl,
         @Override
         public String getNewAuthToken() {
             SalesforceSDKLogger.i(TAG, "Need new access token");
-            final Account acc = clientManager.getAccount();
-            if (acc == null) {
+            UserAccountManager userAccountManager = SalesforceSDKManager.getInstance().getUserAccountManager();
+            Account[] accounts = clientManager.getAccounts();
+            Account matchingAccount = null;
+
+            // Find the account for this client.
+            for (Account account : accounts) {
+                UserAccount user = userAccountManager.buildUserAccount(account);
+                if (user != null && lastNewAuthToken.equals(user.getAuthToken())) {
+                    matchingAccount = account;
+                }
+            }
```

Comment:
I wish there was a better way to do this.  Perhaps we should associate each ClientManager to an account?

Reply:
But don't we have an account associated already with ClientManager: `final Account acc = clientManager.getAccount();`

Reply:
That returns the current account, which is (part of) the problem. If the user we are trying to get a new token for not the current user we will invalidate and refresh for the wrong account.